std/range/primitives.d

```diff
@@ -161,17 +161,11 @@ Params:
 Returns:
     true if R is an InputRange, false if not
  */
-template isInputRange(R)
-{
-    enum bool isInputRange = is(typeof(
-    (inout int = 0)
-    {
-        R r = R.init;     // can define a range object
-        if (r.empty) {}   // can test for empty
-        r.popFront;       // can invoke popFront()
-        auto h = r.front; // can get the front of the range
-    }));
-}
+enum bool isInputRange(R) =
+    is(typeof((R r) => R.init))
```

Comment:
> The only types rejected by this would be those that play shenanigans with init, i.e. have it return a different type etc. We don't want to support such styles. I'll remove the line.

I thought that we were going to make it illegal to redefine `init` (certainly, changes were made to `TypeInfo` so that we could move towards doing that). Until we do, we do need to be a bit careful, but allowing it to be redefined just seems like a disaster waiting to happen - especially when you consider code like this.